libkbfs/md_ops.go

```diff
@@ -353,19 +354,76 @@ func (md *MDOpsStandard) processRange(ctx context.Context, id TlfID,
 		return nil, nil
 	}
 
-	// Verify that the given MD objects form a valid sequence.
-	var prevIRMD ImmutableRootMetadata
-	irmds := make([]ImmutableRootMetadata, 0, len(rmdses))
-	for _, r := range rmdses {
-		bareHandle, err := r.MD.MakeBareTlfHandle()
-		if err != nil {
-			return nil, err
-		}
-		handle, err := MakeTlfHandle(ctx, bareHandle, md.config.KBPKI())
-		if err != nil {
-			return nil, err
+	var wg sync.WaitGroup
+	numWorkers := len(rmdses)
+	if numWorkers > maxMDsAtATime {
+		numWorkers = maxMDsAtATime
+	}
+	wg.Add(numWorkers)
+
+	// Parallelize the MD decryption, because it could involve
+	// fetching blocks to get unembedded block changes.
+	rmdsChan := make(chan *RootMetadataSigned, len(rmdses))
+	irmdChan := make(chan ImmutableRootMetadata, len(rmdses))
+	errChan := make(chan error, 1)
+	worker := func() {
+		defer wg.Done()
+		for rmds := range rmdsChan {
+			bareHandle, err := rmds.MD.MakeBareTlfHandle()
+			if err != nil {
+				select {
+				case errChan <- err:
+				default:
+				}
+				return
+			}
+			handle, err := MakeTlfHandle(ctx, bareHandle, md.config.KBPKI())
+			if err != nil {
+				select {
+				case errChan <- err:
+				default:
+				}
+				return
+			}
+			irmd, err := md.processMetadataWithID(ctx, id, bid, handle, rmds)
+			if err != nil {
+				select {
+				case errChan <- err:
+				default:
+				}
+				return
+			}
+			irmdChan <- irmd
 		}
+	}
+
+	for i := 0; i < numWorkers; i++ {
+		go worker()
+	}
+	for _, rmds := range rmdses {
+		rmdsChan <- rmds
+	}
+	close(rmdsChan)
+	go func() {
+		wg.Wait()
+		close(errChan)
+		close(irmdChan)
+	}()
+	err := <-errChan
+	if err != nil {
+		return nil, err
+	}
+
+	// Sort into slice based on revision.
+	irmds := make([]ImmutableRootMetadata, len(rmdses))
+	for irmd := range irmdChan {
+		irmds[irmd.Revision-rmdses[0].MD.Revision] = irmd
```

Comment:
You might want to check the index here, and return an error if its out of bounds instead of panicking, since I don't think we should necessarily trust that the MDServer returns the RMDs we asked for.